controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,430 @@
+	// If host is not available still create the hfs and can get the ID on subsequent updates
+	info.hfs = hfs
+	if err = r.getProvisionInfoFromBMH(info); err != nil {
+		info.log.Info("could not get provisioner ID", "error", err)
+	}
+
+	if err = r.Create(context.TODO(), hfs); err != nil {
+		return errors.Wrap(err, "could not create hostFirmwareSettings resource")
+	}
+	info.log.Info("created new hostFirmwareSettings resource")
+
+	return nil
+}
+
+// For the baremetalhost that matches the name/namespace get the provisioner ID and hardware info
+func (r *HostFirmwareSettingsReconciler) getProvisionInfoFromBMH(info *rInfo) (err error) {
+
+	bmh := &metal3v1alpha1.BareMetalHost{}
+	if err = r.Get(context.TODO(), info.namespaceName, bmh); err == nil {
+
+		// Store info in annotation that will persist if settings is cleared during pivot
+		info.hfs.Annotations[metal3v1alpha1.ProvisionerIdAnnotation] = bmh.Status.Provisioning.ID
+		// Hardware info is optional
+		if bmh.Status.HardwareDetails != nil {
+			info.hfs.Annotations[metal3v1alpha1.HardwareVendorAnnotation] = bmh.Status.HardwareDetails.SystemVendor.Manufacturer
+			info.hfs.Annotations[metal3v1alpha1.HardwareModelAnnotation] = bmh.Status.HardwareDetails.SystemVendor.ProductName
+		}
+	} else {
+
+		return errors.Wrap(err, "could not get baremetalhost for provisioner id")
+	}
+
+	return nil
+}
+
+// EventHandler for updates to both the baremetalhost and hostfirmwarettings
+func (r *HostFirmwareSettingsReconciler) updateEventHandler(e event.UpdateEvent) bool {
+
+	r.Log.Info("hostfirmwaresettings in event handler")
+
+	// If this is a baremetalhost only return true on certain state transitions
+	var oldState metal3v1alpha1.ProvisioningState = metal3v1alpha1.StateNone
+	var newState metal3v1alpha1.ProvisioningState = metal3v1alpha1.StateNone
+	oldHost, oldHostExists := e.ObjectOld.(*metal3v1alpha1.BareMetalHost)
+	if oldHostExists {
+		oldState = oldHost.Status.Provisioning.State
+	}
+	newHost, newHostExists := e.ObjectNew.(*metal3v1alpha1.BareMetalHost)
+	if newHostExists {
+		newState = newHost.Status.Provisioning.State
+	}
+	if oldHostExists || newHostExists {
+		// The data only needs to be retrieved from Ironic after the node has been moved to
+		// ready and when the node is fully provisionied, its not necessary on every state
+		// change as the data from Ironic will be the same.
+		if newState == metal3v1alpha1.StateReady && oldState == metal3v1alpha1.StatePreparing {
+			r.Log.Info("baremetalhost event state update", "oldstate", oldState, "newstate", newState)
+			return true
+		} else if newState == metal3v1alpha1.StateProvisioned && oldState != metal3v1alpha1.StateProvisioned {
+			r.Log.Info("baremetalhost event state update", "oldstate", oldState, "newstate", newState)
+			return true
+		}
+
+		return false
+	}
+
+	// This is a hostfirmwaresettings resource, if the update increased the resource generation then process it
```

Comment:
Just as an extra safety, an explicit check on the hostfirmwaresettings type could prevent unwanted errors in case a new type will be added in the watch list

Reply:
Added check to confirm its a hostfirmwaresettings resource.